llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -184,17 +185,99 @@ bool transformAccessChain(Function *F) {
   }
   return ModuleChanged;
 }
+
+StoreInst *findLastStoreBeforeLoad(Value *Ptr, Instruction *Load) {
+  BasicBlock::iterator It(Load);
+  while (It != Load->getParent()->begin()) {
+    --It;
+    if (auto *Store = dyn_cast<StoreInst>(&*It))
+      if (Store->getPointerOperand() == Ptr)
+        return Store;
+  }
+  return nullptr;
+}
+
+// Per SPIR-V specification Layout of a matrix must be a constant instruction
+// aka a constexpr or specialization constant. Meanwhile in SYCL headers
+// layout is passed as a parameter to joint_matrix_load function, so even if
+// that layout is a constant expression in the user's code - it's not possible
+// to prove that to the compiler, so constant propagation will happen only
+// after inlining, not in AST. That means, that with O0 layout would remain
+// to be a runtime variable in LLVM IR.
+// SYCL matrix layout is being mapped on SPIR-V matrix layout by
+// joint_matrix_layout_to_spv function. The following routine finds calls to
+// this function and replaces them with the found constant.
+// This function also cleans up code, that becomes dead. Pattern of the dead
+// code is stable, as user's code doesn't affect it.
+bool propagateConstexprLayout(Function *F) {
+  llvm::SmallVector<Instruction *, 8> ToErase;
+  for (auto I = F->user_begin(), E = F->user_end(); I != E;) {
+    User *U = *I++;
+    auto *CI = dyn_cast<CallInst>(U);
+    if (!CI)
+      continue;
+    auto *Op = dyn_cast<Instruction>(CI->getArgOperand(0));
+    if (!Op || !isa<LoadInst>(Op))
+      continue;
+    auto *Ptr = dyn_cast<Instruction>(cast<LoadInst>(Op)->getPointerOperand());
+    if (!Ptr)
+      continue;
+
+    ConstantInt *ConstLayout = nullptr;
+    StoreInst *SI = findLastStoreBeforeLoad(Ptr, Op);
+    if (!SI)
+      continue;
+    ConstLayout = dyn_cast<ConstantInt>(SI->getValueOperand());
+    if (ConstLayout) {
+      CI->replaceAllUsesWith(ConstLayout);
+      ToErase.push_back(CI);
+      ToErase.push_back(SI);
+      ToErase.push_back(Op);
+      ToErase.push_back(Ptr);
+      if (auto *Cast = dyn_cast<AddrSpaceCastInst>(Ptr)) {
+        auto *OrigPtr = Cast->getPointerOperand();
+        if (auto *AI = dyn_cast<AllocaInst>(OrigPtr))
+          ToErase.push_back(AI);
+      }
+    }
+  }
+
+  // There are possible cases, when a single instruction result is used multiple
+  // times. For this case we have to use a vector to store such instructions
+  // and keep track if we have removed them before to avoid double free().
+  SmallPtrSet<Instruction *, 8> Erased;
```

Comment:
It looks a bit awkward, but I haven't figured out a better way. If I would use `SmallPtrSet` to store `ToErase` instructions, then it would result in instructions not being removed in several scenarios, when at the point, when they were processed number of users is still not zero.

Reply:
Just a thought. Up to you...

Would it be better to structure algorithm that way?
1. Populate `ToErase` with just `CI`, `Op` and `SI`, populate separate set with `Ptr` and `AI`.
2. Erase `ToErase`
3. Then erase `Ptr` and `AI`

Reply:
It should work. But I don't have an opinion, which is less awkward :)
Still I need to have a review from either @AlexeySachkov or @intel/dpcpp-tools-reviewers , may be they will provide an opinion, which way is better.

Reply:
You can try something like this.
if (!II->use_empty())
II->replaceAllUsesWith(UndefValue::get(II->getType()));
and then remove II

Reply:
I'm not sure, how it would help to avoid double freeing the memory, we still have to either keep track of what we have removed and/or replaced, haven't we?

Reply:
Ah...I see now..'ToErase' is a vector and can have duplicate entries. Why not make it a set?

Thanks

Reply:
> Why not make it a set?

In this case another problem shows up: 
lets say we have a single alloca and then 2 loads from it and 2 calls to `joint_matrix_layout_to_spv` function. Order we add instructions to the set:
1st call to joint_matrix_layout_to_spv -> 1st load from the alloca -> alloca -> 2nd call to joint_matrix_layout_to_spv -> 2nd load from the alloca -> alloca (not added, since it's a set).

This means that with a set we won't remove the alloca, since it's used by the 2nd load (aka !use_empty()). So in this case we would need to go over the set multiple times.

So the vector is used - we can store a single instruction multiple times if it's used multiple times. But we had to keep track of the removed instructions to avoid double free.

Reply:
Hmm. I think Yury's suggestion seems a bit cleaner. But, I am ok with existing code as well.

Thanks